webrender/src/tiling.rs

```diff
@@ -2663,10 +2600,58 @@ impl FrameBuilder {
                                                                     device_pixel_ratio);
                             }
+
+                            // If the primitive is visible, consider culling it via clip rect(s).
+                            // If it is visible but has clips, create the clip task for it.
+                            if let Some(prim_bounding_rect) = self.prim_store
+                                                                  .cpu_bounding_rects[prim_index.0] {
+                                let prim_metadata = &mut self.prim_store.cpu_metadata[prim_index.0];
+                                let prim_clip_info = prim_metadata.clip_cache_info.as_ref();
+                                let mut visible = true;
+
+                                if let Some(info) = prim_clip_info {
+                                    clip_info_stack.push((*sc_index, info.clone()));
+                                }
+
+                                // Try to create a mask if we may need to.
+                                if prim_clip_info.is_some() || !clip_info_stack.is_empty() {
+                                    let mask_opt = RenderTask::new_mask(prim_bounding_rect,
+                                                                        MaskCacheKey::Primitive(prim_index),
+                                                                        &clip_info_stack,
+                                                                        &self.layer_store);
+                                    match mask_opt {
+                                        MaskResult::Outside => {
+                                            // Primitive is completely clipped out.
+                                            prim_metadata.clip_task = None;
+                                            visible = false;
+                                        }
+                                        MaskResult::Inside(task) => {
+                                            // Got a valid clip task, so store it for this primitive.
+                                            prim_metadata.clip_task = Some(task);
+                                        }
+                                    }
+                                }
+
+                                if let Some(..) = prim_clip_info {
+                                    clip_info_stack.pop();
+                                }
+
+                                if visible {
+                                    profile_counters.visible_primitives.inc();
+                                }
+                            }
                         }
                     }
                 }
                 &PrimitiveRunCmd::PopStackingContext => {
+                    let sc_index = *layer_stack.last().unwrap();
+                    let layer = &mut self.layer_store[sc_index.0];
+                    if layer.is_visible() {
```

Comment:
the push stuff is using a different check - `can_contribute_to_scene()`, so there might be a bug here